### PR TITLE
Display "Add review" link if there's no product rating

### LINKED ISSUE #7923: Display "Add review" link if there's no product rating
### Description
If product has no ratings, "Add rating" link appears instead of stars. For now the link text is NOT modifiable and redirects to #reviews section. Change applies to:

- All Products
- Products
- Legacy PHP template

### Gain
It would help with preserving the layout and eventually be a consistent way to link shoppers directly to the rating results for the product.

---

Note: The scope of this issue has changed based on the discussion in https://github.com/woocommerce/woocommerce-blocks/pull/7929

### PR DESCRIPTION
**Pull request by kmanijak**
#### This PR is repurposed based on the discussion in the comments.

TLDR:
Previously: Add 5 empty stars when there's no rating
Currently: Add "Add review" link to the products with no ratings
Reason for change: "Empty" stars could be interpreted as rating 0/5, which could give a false impression

--- 

#### Description

It was suggested [within the scope of Product Query patterns ](https://github.com/woocommerce/woocommerce-blocks/pull/7857#issuecomment-1346386292) to display something when there's no product ratings, so the layout is preserved.

This PR enables "Add review" link in:
- All Products
- Products
in case there's no rating.

"Add review" link redirects to `<<product_URL>>#reviews`, so to the reviews section.

Previously there was no "loading" placeholder, since it wasn't determined if there will be any content rendered. Currently, as it's determined that either rating or a link will be rendered, the placeholder is visible during loading.

Fixes #7923 

#### Accessibility

- [x] I've tested using only a keyboard (no mouse)
- [x] I've tested using a screen reader

### Screenshots

| Case | Before | After |
| ------ | ----- |----- |
| Products (Editor) |  image   |    image  |
| Products (Frontend) |  image   |   image   |
| All Products (Editor) |  image   |  image    |
| All Products (Frontend) |  image   |   image   |

#### User Facing Testing

Prerequisites:
- make sure you have at least one product **with** and at least one product **without** rating

##### All Products and Products blocks

Steps:
1. Go to Editor and add _All Products_ block
2. Make sure Rating is included in product layout (Pencil icon > Use inserter > Add "Product Rating" block)
3. Save the template

**Expected:**
- Both products, _with_ and _without_ rating, have the same layout.
- Products with no rating have "Add review" link
- Link is not clickable in Editor

4. Go to the Frontend

**Expected:**
- Both products, _with_ and _without_ rating, have the same layout.
- Products with no rating have "Add review" link
- Link is interactive and redirects to Review section of the product

**Repeat the same for the Products block!**

##### Regression testing - Product view
1. Go to product page of the product without rating.

**Expected:**
- There's no "Add review" link instead or rating stars

##### Regression testing - Rating alignment

1. Check the Alignment settings in both: _Products_ for both: stars and link!
image

**Expected:**
- Alignment is respected in Editor and Frontend for all: left, center and right alignment

##### Regression testing - Filer by Rating

1. Add Filter by Rating block
2. Save and go to Frontend

**Expected:**
- There's no "loading" placeholder in place of stars in the Filter by Rating

### WooCommerce Visibility

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Changelog

> Products and All Products: Display "Add review" link when there's no product rating

### CODE REVIEW

**Comment by nerrad** [comment]
Thanks for doing this! I see in your screenshots that the price for some products is wrapped on inline with the ratings? Do you know if this is intentional? I think ratings and price should be stacked right?

**Comment by Aljullu** [comment]
I didn't review it yet, but I'm a bit concerned that the empty stars look like a 0 rating. IMO the _Before_ behavior shown in the screenshots is a valid one and can be seen online in existing stores, so I do think we should support it.

Some thoughts:

* Could this be made an option so stores can decide which behavior they prefer?
* If the issue is about the layout breaking when some products have ratings and others don't. Could we explore adding a _Minimum height_ to the _Product rating_ block? That would preserve the layout in cases where some products have ratings and others don't. (Sidenote: if we go this way, I'm uncertain whether it's better to use the _Minimum height_ global styles option or hard-coding it in CSS)
* Another idea would be to have a different layout for 'no rating' stars. Ie, something like only the outline of the stars. Or simply displaying a 'No ratings' text.

@nerrad

**Comment by danieldudzic** [comment]
> I didn't review it yet, but I'm a bit concerned that the empty stars look like a 0 rating. IMO the _Before_ behavior shown in the screenshots is a valid one and can be seen online in existing stores, so I do think we should support it.

>   Another idea would be to have a different layout for 'no rating' stars. Ie, something like only the outline of the stars. Or simply displaying a 'No ratings' text.

In my opinion, there should be an option to hide empty ratings, but I also second the idea to use a different design for no ratings - the outline sounds reasonable and doesn't clutter the layout with extraneous text.

**Comment by nerrad** [comment]
> I didn't review it yet, but I'm a bit concerned that the empty stars look like a 0 rating.

I didn't think of that but it makes sense.

> Could this be made an option so stores can decide which behavior they prefer?

Karol asked me this in chat and I initially answered we should wait until we get feedback but if it's trivial to add the toggle I'm okay with doing so. I'm not sure we should go with an alternative design to represent "no ratings", I'm not sure how that's any more clearer than the current iteration. I imagine "outline" vs "fill" for ratings might be a design choice as well for the stars themselves eventually.

> Sidenote: if we go this way, I'm uncertain whether it's better to use the Minimum height global styles option or hard-coding it in CSS

If adding an option to hide no ratings is more complex, happy to try this approach first and agree the minimum height global styles option is likely the best approach here (I'm not sure what the rendered CSS is offhand though and if there are additional complexities we need to consider about potential surrounding blocks in patterns etc).

**Comment by kmanijak** [comment]
> I see in your screenshots that the price for some products is wrapped on inline with the ratings? Do you know if this is intentional? I think ratings and price should be stacked right?

It happens for PHP template, and I assumed it's expected since that's how it works on `trunk` as well. But once you mentioned that, I'm pretty sure it's not expected. Let me create an issue. 👍 

---
> if it's trivial to add the toggle I'm okay with doing so

> If adding an option to hide no ratings is more complex, happy to try this approach first and agree the minimum height global styles option is likely the best approach here 

So in case of adding a toggle that would mean:
- Add toggle to All Products settings: "Show empty ratings" with default value `false`
- Add toggle to Product Ratings block settings (used in Products): "Show empty ratings" with default value `false`
- To my understanding, PHP template is not configurable. In such a case shall we always show or always hide the no rating?

It's hard for me to assess if it's trivial or not. I can timebox the toggle work to 30-60minutes and in case it's becoming complex, I'll switch to min-height solution. How does it sound?

**Comment by nerrad** [comment]
> Add toggle to All Products settings: "Show empty ratings" with default value false
> Add toggle to Product Ratings block settings (used in Products): "Show empty ratings" with default value false

Wouldn't it just be a control configured with the Product Ratings block (and thus available in both contexts)? The products rating block can be selected with the All Products block. 

> To my understanding, PHP template is not configurable. In such a case shall we always show or always hide the no rating?

If you're talking about the rendered classic template, correct - it's not configurable and we won't worry about the classic templates for now. As we blockify the templates the existing product element blocks will allow for finer control.

**Comment by kmanijak** [comment]
> Wouldn't it just be a control configured with the Product Ratings block (and thus available in both contexts)? The products rating block can be selected with the All Products block.

My bad, you're right!

But I'm thinking now: the main reason for the change was to preserve the layout. Making the empty rating configurable with the default to NOT show the empty rating, I'd definitely go with min-height approach.
In this case empty rating helps only those who enable that.
Min-height would provide generic solution for everyone.

**Comment by nefeline** [comment]
> Another idea would be to have a different layout for 'no rating' stars. Ie, something like only the outline of the stars. Or simply displaying a 'No ratings' text.

I like this idea too. Sharing my 1% here:

I've seen websites where there's text on the side of the star ratings indicating when a product has no reviews:

Screenshot 2022-12-13 at 19 08 24

When reviews are available, the label is then updated to:

Screenshot 2022-12-13 at 19 08 42


Maybe that's an option we can explore?

**Comment by kmanijak** [comment]
@nefeline, I like the idea, but I'm worried that adding text next to the stars may make the UI too crowded, especially on product grids.

---

I went through couple of online shops and another interesting idea is to either display a star rating or "Add review" link which brings you to `#reviews` section . What do you think about that?

image

**Comment by nefeline** [comment]
> I like the idea, but I'm worried that adding text next to the stars may make the UI too crowded, especially on product grids.

That's a fair concern

> I went through couple of online shops and another interesting idea is to either display a star rating or "Add review" link which brings you to #reviews section . What do you think about that?

My 2 cents: I like this approach, it seems to be very clean and intuitive info to whoever is accessing the store.

**Comment by Aljullu** [comment]
> I went through couple of online shops and another interesting idea is to either display a star rating or "Add review" link which brings you to #reviews section . What do you think about that?

I like this approach as well and prefer it over the "empty stars". In an ideal scenario, it would be possible to customize the text, so stores could delete it if they want that space to be empty.

Something similar to the _Read more_ text in the _Product Summary_ block:

Product Summary Read more text screenshot

**Comment by nerrad** [comment]
> I went through couple of online shops and another interesting idea is to either display a star rating or "Add review" link which brings you to #reviews section . What do you think about that?

I like this approach too! Agree with Albert that future iterations could provide an option for customizing but for now, we could just have the default.

**Comment by kmanijak** [comment]
> Agree with Albert that future iterations could provide an option for customizing but for now, we could just have the default.

Great!

So to sum up. If product has no ratings, "Add rating" link appears instead of stars. For now the link text is NOT modifiable and redirects to #reviews section. Change applies to:
- All Products
- Products
- Legacy PHP template

I'll go ahead and reflect the outcome of this discussion in the issue https://github.com/woocommerce/woocommerce-blocks/issues/7923 and will repurpose this PR.

Thanks all for a great discussion! 🙌

**Comment by kmanijak** [comment]
@nefeline , @danieldudzic , PR has been updated and is ready for a review 🙏

**Review by danieldudzic — COMMENTED** [review]
> Products and All Products: Display "Add review" link when there's no product rating

Maybe I'm nitpicking here, but perhaps it would make sense to indicate in the changelog message that this PR is relevant to the Rating product element:

> Rating: Display "Add review" link when there's no product rating.

File: src/BlockTypes/ProductRating.php [hunk 1]
```diff
@@ -113,14 +113,22 @@ protected function render( $attributes, $content, $block ) {
 			$styles_and_classes            = StyleAttributesUtils::get_classes_and_styles_by_attributes( $attributes, array( 'font_size', 'margin', 'text_color' ) );
 			$text_align_styles_and_classes = StyleAttributesUtils::get_text_align_class_and_style( $attributes );
 
+			$average_rating = $product->get_average_rating();
+			$rating_count   = $product->get_rating_count();
+			$reviews_link   = $product->get_permalink() . '#reviews';
+			$link_label     = __( 'Add review', 'woo-gutenberg-products-block' );
+			$link_html      = '<a class="wc-block-components-product-rating__link" href="' . $reviews_link . '">' . esc_attr( $link_label ) . '</a>';
```
Comment (danieldudzic) [inline src/BlockTypes/ProductRating.php]: It might make sense to escape the `href` here:
```suggestion
			$link_html      = '<a class="wc-block-components-product-rating__link" href="' . esc_url( $reviews_link ) . '">' . esc_attr( $link_label ) . '</a>';;
```

File: assets/js/atomic/blocks/product-elements/rating/style.scss [hunk 2]
```diff
@@ -45,9 +45,26 @@
 			white-space: nowrap;
 		}
 	}
+
+	&__link {
+		display: inline-block;
+		width: 100%;
+		text-align: center;
```
Comment (danieldudzic) [inline assets/js/atomic/blocks/product-elements/rating/style.scss]: This is breaking the `Alignment` block setting for the link.
```suggestion

```

**Review by danieldudzic — CHANGES_REQUESTED** [review]
The PR is working well and adding the `Add Review` link correctly. Good job!

I have noticed:

1. The `Alignment` block setting is currently broken for the `Rating` element. I have tracked it down to this commit: https://github.com/woocommerce/woocommerce-blocks/commit/cfb920a0758613fcdb61c9a455f1516a55ef13b3

2. The `Add review` link should probably inherit the custom color setting:

Edit_Page_“Products”_‹_ratings_—_WordPress

File: src/BlockTypes/ProductRating.php [hunk 3]
```diff
@@ -113,14 +113,22 @@ protected function render( $attributes, $content, $block ) {
...
+
 			return sprintf(
 				'<div class="wc-block-components-product-rating wc-block-grid__product-rating %1$s %2$s" style="%3$s">
 					%4$s
 				</div>',
 				esc_attr( $text_align_styles_and_classes['class'] ?? '' ),
 				esc_attr( $styles_and_classes['classes'] ),
 				esc_attr( $styles_and_classes['styles'] ?? '' ),
-				wc_get_rating_html( $product->get_average_rating() )
+				$content
```
Comment (nefeline) [inline src/BlockTypes/ProductRating.php]: Since we are already adding a filter to modify WooCommerce's default rating HTML here on this Class (see the `filter_rating_html` method on `L78`), my recommendation is to move and consolidate the updates done here on `render` to that method instead. E.g.:

```php
	public function filter_rating_html( $html, $rating, $count ) {
		$product_permalink = get_permalink();
		if ( 0 < $rating || false === $product_permalink ) {
			/* translators: %s: rating */
			$label   = sprintf( __( 'Rated %s out of 5', 'woo-gutenberg-products-block' ), $rating );
			$content = '<div class="wc-block-components-product-rating__stars wc-block-grid__product-rating__stars" role="img" aria-label="' . esc_attr( $label ) . '">' . wc_get_star_rating_html( $rating, $count ) . '</div>';
		} else {
			$product_review_url = esc_url( $product_permalink . '#reviews' );
			$content            = '<a class="wc-block-components-product-rating__link" href="' . $product_review_url . '">' . __( 'Add review', 'woo-gutenberg-products-block' ) . '</a>';
		}

		return $content;
	}
```

**Review by nefeline — CHANGES_REQUESTED** [review]
Works like a charm 👏 ! This is testing well: also noticed the same points that @danieldudzic already flagged [here](https://github.com/woocommerce/woocommerce-blocks/pull/7929/#pullrequestreview-1220266658). 

I've recommended a change regarding how the rating HTML is structured since it is already being modified on the `filter_rating_html` method.

**Comment by kmanijak** [comment]
@danieldudzic and @nefeline , I appreciate the reviews! 🙏 

### Review

> The Alignment block setting is currently broken for the Rating element. I have tracked it down to this commit: https://github.com/woocommerce/woocommerce-blocks/commit/cfb920a0758613fcdb61c9a455f1516a55ef13b3

You're right, I fixed that in this PR

> The Add review link should probably inherit the custom color setting:

It does, but for links, which it is... But that brings inconsistency: stars (which aren't links) follow the text color, while "Add review" follow the link color.

image

To be honest, not sure what's better: semantic consistency (as it is right now) or blocks consistency. Do you have any opinion?

### Remaining work 

There's still one remaining thing I need to address: I added a placeholder while the Rating is loading. However that influences the Filter by Rating block (by accident it has `is-loading` class all the time).

File: src/BlockTypes/ProductRating.php (hunk 3 above)
Comment (kmanijak) [reply src/BlockTypes/ProductRating.php]: First of all: thank you for this, that was my first idea, but I realised I have no clue where to get the `product_permalink` from. I didn't know I can simply call `get_permalink()` and that will work in a context of a particular product, so... Today I Learnt, thank you! 🎉 
 
But once I tried that I found a bad consequences. This filter influences not only Products block, but also Classic Template. I'm not sure why and where's the mechanism responsible for that, but when another anchor is put inside (which "Add review" is), the Rating and a Price are pushed out of the product anchor. That brings inconsistency. Check the image out:
- on the left: product which is wrapped in anchor HTML tag. Everything has a green color, since that's how I configured links
- on the right: "Add review" is a link and it's been pushed out of product anchor. Price is red, which is a color of regular text.

image

Given the above I think I would go back to the previous solution. What do you think?

**Comment by kmanijak** [comment]
@nefeline , @danieldudzic 

I updated the PR with the fix for loading placeholder issue which was introducing regression in Filter by Rating. In short: there's always `is-loading` class attached in the Filter by Rating. I wanted to fix that, however there's couple of edge cases that would have to be covered, hence I decided to create an issue for that. For it's just fixed by limiting the set of classes to which placeholder is applied ([code](https://github.com/woocommerce/woocommerce-blocks/pull/7929/files#diff-0ec89a253e28931bdfdbb528c1b9f187e66dad78d2b325059234e94075294732R64-R74)).

I added few regression testing steps to the PR description, for Product view, Rating alignment and Filter by Rating.

Also, I responded to some of your comments with my concerns, I'm happy to hear your opinion about them. Other comments, with 👍 reaction are addressed.

I'd be grateful for another round of review here 🙏

File: src/BlockTypes/ProductRating.php (hunk 3 above)
Comment (nefeline) [reply src/BlockTypes/ProductRating.php]: > First of all: thank you for this, that was my first idea, but I realised I have no clue where to get the product_permalink from. I didn't know I can simply call get_permalink() and that will work in a context of a particular product, so... Today I Learnt, thank you! 🎉

Anytime 🙌 !

> This filter influences not only Products block, but also Classic Template.

Out of curiosity: why would we want to display the "add review" link only on blocks?

> I'm not sure why and where's the mechanism responsible for that, but when another anchor is put inside (which "Add review" is), the Rating and a Price are pushed out of the product anchor. That brings inconsistency.

I'm probably missing a piece of information here: do you mind clarifying this a bit more? Isn't this a consequence of what is being discussed [over here](https://github.com/woocommerce/woocommerce-blocks/pull/7929#issuecomment-1357748183) (stars (which aren't links) follow the text color, while "Add review" follow the link color.)? I've compared both implementations side by side (consolidating everything on `filter_rating_html` versus keeping the changes on `render`), and I do not see any differences in terms of the structure of the divs or HTML anchors.

> Given the above I think I would go back to the previous solution. What do you think?

No strong preferences on my end, although filtering the rating_html and then modifying the filtered result right after, on the same class, looks a bit redundant/confusing.

Comment (kmanijak) [reply src/BlockTypes/ProductRating.php]: Thanks @nefeline for response!

> Out of curiosity: why would we want to display the "add review" link only on blocks?

It's not like we want the link only on blocks. Actually it's a good step to cover Classic Template as well, but there's no high demand for that.

> I'm probably missing a piece of information here: do you mind clarifying this a bit more? 

Sure thing! Apologies it wasn't clear enough.

On the screen below you can see the that when there are stars (green), they are rendered within the product anchor (red) and the same goes for price. Product anchor wraps the whole content:
Screen Shot 2022-12-22 at 18 21 58 PM

However, when "Add Review" link is rendered, the Product anchor gets closed (I think browser may sanitize that) to avoid nested anchors (which is forbidden). The consequence is that "Add Review" link (green) and price are already outside of Product anchor (red). That changes the clicking area for product and causes potential UI differences if a user changed the global styles. It cannot be easily fixed as Classic Template "delegates" single product rendering to WC Core.
Screen Shot 2022-12-22 at 18 21 31 PM

Having said that, I think we should skip the Classic Template. Let me know if you have some idea how to overcome this, because I have to agree with your words:

> filtering the rating_html and then modifying the filtered result right after, on the same class, looks a bit redundant/confusing.

Comment (kmanijak) [reply src/BlockTypes/ProductRating.php]: Thanks @nefeline for the effort you took to show me why the above case is an edge case. It only happens when both: Products block and Classic Template are used on the same page (which I did to make the testing easier and didn't understand it influences the result).

When Classic Template is used alone on the page it stays untouched. It's very unlikely a merchant would use both those components on the same page, hence I applied the change as per your recommendation.

Having said that I think PR is ready for a final check! 🙌

Comment (nefeline) [reply src/BlockTypes/ProductRating.php]: Thanks for the update here @kmanijak ! This looks good now that we have removed the redundancy of modifying an already-filtered result.

> It only happens when both: Products block and Classic Template are used on the same page

Not at all related to this PR in particular since the filter is pre-existing on the `ProductRating` class, but to prevent this behavior altogether and keep the Product Rating block customizations isolated, we would have to stop relying on the `'woocommerce_product_get_rating_html'` filter and work ourselves on a parallel implementation instead.

> It's very unlikely a merchant would use both those components on the same page

Yes, I agree: and even if they do, it still renders great and works as expected, except, of course, the custom block styles might not be applied on the Classic Template for obvious reasons :)

**Review by nefeline — APPROVED** [review]
I've confirmed everything works as expected. Great work! 🚀

**Comment by kmanijak** [comment]
Merging with the bypass - E2E tests with Gutenberg fail, the same as on `trunk`. All the other checks are passing, including E2E without Gutenberg.
